File: flexlay/lib/object_layer.cxx
//  $Id$
//
//  Flexlay - A Generic 2D Game Editor
//  Copyright (C) 2002 Ingo Ruhnke <grumbel@gmx.de>
//
//  This program is free software; you can redistribute it and/or
//  modify it under the terms of the GNU General Public License
//  as published by the Free Software Foundation; either version 2
//  of the License, or (at your option) any later version.
//
//  This program is distributed in the hope that it will be useful,
//  but WITHOUT ANY WARRANTY; without even the implied warranty of
//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//  GNU General Public License for more details.
//
//  You should have received a copy of the GNU General Public License
//  along with this program; if not, write to the Free Software
//  Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.

#include <iostream>
#include <ClanLib/Display/display.h>
#include <ClanLib/Core/Math/origin.h>
#include <ClanLib/Core/System/error.h>
#include "objmap_object.hxx"
#include "objmap_sprite_object.hxx"
#include "objmap_control_point.hxx"
#include "object_layer.hxx"
#include "editor_map_component.hxx"
#include "layer_impl.hxx"

ObjectLayer ObjectLayer::current_;

class ObjectLayerImpl : public LayerImpl
{
public:
  ObjectLayer::Objects objects;
  ObjectLayer::ControlPoints control_points;
  CL_SlotContainer slots;

  ObjectLayerImpl() {}
  virtual ~ObjectLayerImpl() {}
  
  void draw(EditorMapComponent* parent, CL_GraphicContext* gc);
  bool has_bounding_rect() const { return false; }
};

ObjectLayer::ObjectLayer()
  : impl(new ObjectLayerImpl())
{
}

ObjectLayer::~ObjectLayer()
{
}

void
ObjectLayerImpl::draw(EditorMapComponent* parent, CL_GraphicContext* gc)
{
  for(ObjectLayer::Objects::iterator i = objects.begin(); i != objects.end(); ++i)
    {
      // FIXME: Add clipping here
      if (parent->get_clip_rect().is_overlapped((*i).get_bound_rect()))
        (*i).draw(gc);
    }

  for(ObjectLayer::ControlPoints::iterator i = control_points.begin(); i != control_points.end(); ++i)
    {
      (*i).draw(gc);
    }
}

ObjMapControlPoint
ObjectLayer::find_control_point(const CL_Pointf& click_pos)
{
  for(ControlPoints::reverse_iterator i = impl->control_points.rbegin(); 
      i != impl->control_points.rend(); 
      ++i)
    {
      CL_Rect rect = (*i).get_bound_rect();
     
      if (rect.is_inside(CL_Point(click_pos)))
        return *i;
    }
  return ObjMapControlPoint(); 
}

ObjMapObject
ObjectLayer::find_object(const CL_Pointf& click_pos)
{
  for(Objects::reverse_iterator i = impl->objects.rbegin(); i != impl->objects.rend(); ++i)
    {
      CL_Rectf rect = (*i).get_bound_rect();
     
      if (rect.is_inside(CL_Point(click_pos)))
        return *i;
    }
  return ObjMapObject();
}

void
ObjectLayer::delete_object(const ObjMapObject& obj)
{
  for(Objects::iterator i = impl->objects.begin(); i != impl->objects.end(); ++i)
    {
      if (obj == (*i))
        {
          impl->objects.erase(i);
          break;
        }
    }
}

ObjectLayer::Objects
ObjectLayer::get_selection(const CL_Rectf& rect)
{
  Objects selection;

  for(Objects::iterator i = impl->objects.begin(); i != impl->objects.end(); ++i)
    {
      // FIXME:
      if (rect.is_inside((*i).get_pos()))
        {
          selection.push_back(*i);
        }
    }
  
  return selection;
}

ObjectLayer::Objects
ObjectLayer::get_objects()
{
  return impl->objects;
}

void
ObjectLayer::add_object(const ObjMapObject& obj)
{
  impl->objects.push_back(obj);
}

void
ObjectLayer::add_control_point(const ObjMapControlPoint& obj)
{
  impl->control_points.push_back(obj);
}

void
ObjectLayer::delete_control_points()
{
  impl->control_points.clear();
}

Layer
ObjectLayer::to_layer()
{
  return Layer(impl);
}

int
ObjectLayer::get_object_index(const ObjMapObject& obj)
{
  Objects::size_type i;
  for(i = 0; i < impl->objects.size(); ++i)
    {
      if (impl->objects[i] == obj)
        {
          return i;
        }
    }
  return -1;
}

void
ObjectLayer::move_to(const ObjMapObject& obj, int height)
{
  // FIXME: Implement me
}

void
ObjectLayer::raise(const ObjMapObject& obj)
{
  int i = get_object_index(obj);
  if (i != -1 && impl->objects.size() > 1 && i < int(impl->objects.size())-1)
    {
      std::swap(impl->objects[i], impl->objects[i+1]);
    }
}

void
ObjectLayer::lower(const ObjMapObject& obj)
{
  int i = get_object_index(obj);
  if (i != -1 && i > 0)
    {
      std::swap(impl->objects[i], impl->objects[i-1]);
    }
}

/* EOF */

File: flexlay/object_layer.py
# Flexlay - A Generic 2D Game Editor
# Copyright (C) 2014 Ingo Ruhnke <grumbel@gmail.com>
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.


from typing import Optional

from flexlay.graphic_context import GraphicContext
from flexlay.layer import Layer
from flexlay.math import Pointf, Rectf
from flexlay.objmap_control_point import ObjMapControlPoint
from flexlay.objmap_object import ObjMapObject


class ObjectLayer(Layer):

    def __init__(self) -> None:
        super().__init__()

        self.objects: list[ObjMapObject] = []
        self.control_points: list[ObjMapControlPoint] = []

    def has_bounding_rect(self) -> bool:
        return False

    def draw(self, gc: GraphicContext) -> None:
        for obj in self.objects:
            rect = obj.get_bound_rect()
            if obj.to_draw and (rect is None or gc.get_clip_rect().is_overlapped(rect)):
                obj.draw(gc)

        for cp in self.control_points:
            cp.draw(gc)

    def find_control_point(self, click_pos: Pointf) -> Optional[ObjMapControlPoint]:
        for cp in reversed(self.control_points):
            rect = cp.get_bound_rect()
            if rect is not None and rect.is_inside(click_pos):
                return cp
        return None

    def find_object(self, click_pos: Pointf) -> Optional[ObjMapObject]:
        for obj in reversed(self.objects):
            if obj.is_inside(click_pos):
                return obj
        return None

    def delete_object(self, obj: ObjMapObject) -> None:
        self.objects.remove(obj)

    def delete_objects(self, objs: list[ObjMapObject]) -> None:
        for obj in objs:
            self.objects.remove(obj)

    def get_selection(self, rect: Rectf) -> list[ObjMapObject]:
        selection = []
        for obj in self.objects:
            if rect.is_inside(obj.get_pos()):
                selection.append(obj)

        return selection

    def get_objects(self) -> list[ObjMapObject]:
        return self.objects

    def add_object(self, obj: ObjMapObject) -> None:
        self.objects.append(obj)

    def add_control_point(self, obj: ObjMapControlPoint) -> None:
        self.control_points.append(obj)

    def delete_control_points(self) -> None:
        self.control_points.clear()

    def get_object_index(self, needle: ObjMapObject) -> int:
        for idx, obj in enumerate(self.objects):
            if obj == needle:
                return idx
        return -1

    def raise_objects_to_top(self, objs: list[ObjMapObject]) -> None:
        self.delete_objects(objs)
        self.objects.extend(objs)

    def lower_objects_to_bottom(self, objs: list[ObjMapObject]) -> None:
        self.delete_objects(objs)
        self.objects = objs + self.objects

    def raise_object(self, obj: ObjMapObject) -> None:
        i = self.get_object_index(obj)
        if i != -1 and len(self.objects) > 1 and i < len(self.objects) - 1:
            self.objects[i], self.objects[i + 1] = self.objects[i + 1], self.objects[i]

    def lower_object(self, obj: ObjMapObject) -> None:
        i = self.get_object_index(obj)
        if i != -1 and i > 0:
            self.objects[i], self.objects[i - 1] = self.objects[i - 1], self.objects[i]


# EOF #
